ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionSimulator.java

```diff
@@ -0,0 +1,203 @@
+
+  // Dummy signature for transactions to not fail being processed.
+  private static final SECP256K1.Signature FAKE_SIGNATURE =
+      SECP256K1.Signature.create(SECP256K1.HALF_CURVE_ORDER, SECP256K1.HALF_CURVE_ORDER, (byte) 0);
+
+  private static final Address DEFAULT_FROM =
+      Address.fromHexString("0x0000000000000000000000000000000000000000");
+
+  private final Blockchain blockchain;
+  private final WorldStateArchive worldStateArchive;
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final PrivacyParameters privacyParameters;
+
+  public PrivateTransactionSimulator(
+      final Blockchain blockchain,
+      final WorldStateArchive worldStateArchive,
+      final ProtocolSchedule<?> protocolSchedule,
+      final PrivacyParameters privacyParameters) {
+    this.blockchain = blockchain;
+    this.worldStateArchive = worldStateArchive;
+    this.protocolSchedule = protocolSchedule;
+    this.privacyParameters = privacyParameters;
+  }
+
+  public Optional<PrivateTransactionProcessor.Result> process(
+      final String privacyGroupId,
+      final String enclaveKey,
+      final CallParameter callParams,
+      final Hash blockHeaderHash) {
+    final BlockHeader header = blockchain.getBlockHeader(blockHeaderHash).orElse(null);
+    return process(privacyGroupId, enclaveKey, callParams, header);
+  }
+
+  public Optional<PrivateTransactionProcessor.Result> process(
+      final String privacyGroupId,
+      final String enclaveKey,
+      final CallParameter callParams,
+      final long blockNumber) {
+    final BlockHeader header = blockchain.getBlockHeader(blockNumber).orElse(null);
+    return process(privacyGroupId, enclaveKey, callParams, header);
+  }
+
+  public Optional<PrivateTransactionProcessor.Result> processAtHead(
+      final String privacyGroupId, final String enclaveKey, final CallParameter callParams) {
+    return process(privacyGroupId, enclaveKey, callParams, blockchain.getChainHeadHeader());
+  }
+
+  private Optional<PrivateTransactionProcessor.Result> process(
+      final String privacyGroupIdString,
+      final String enclaveKey,
+      final CallParameter callParams,
+      final BlockHeader header) {
+    if (header == null) {
+      return Optional.empty();
+    }
+
+    // check whether the user (enclaveKey) is part of the current privacy group
+    PrivacyGroup privacyGroup = null;
+    try {
+      privacyGroup = privacyParameters.getEnclave().retrievePrivacyGroup(privacyGroupIdString);
+    } catch (final EnclaveClientException e) {
+      // do not do anything. This is handled in the following if
+    }
+    if (privacyGroup == null || !privacyGroup.getMembers().contains(enclaveKey)) {
+      return Optional.of(
+          PrivateTransactionProcessor.Result.failed(
+              0L,
+              ValidationResult.invalid(
+                  TransactionValidator.TransactionInvalidReason.PRIVACY_GROUP_DOES_NOT_EXIST),
+              Optional.empty()));
+    }
```

Comment:
NIT: This method is quite big. It would be good if we could wrap some of the logic into smaller private methods to keep the main method cleaner.

Example:

```
//refactored
if (!nodeHasAccessToPrivacyGroup(privacyGroupIdString, enclaveKey)) {
      return privacyGroupDoesNotExistResult();
}
```
----------------
```
  // priv method 1
  private Optional<Result> privacyGroupDoesNotExistResult() {
    return Optional.of(
        Result.failed(
            0L,
            ValidationResult.invalid(
                TransactionValidator.TransactionInvalidReason.PRIVACY_GROUP_DOES_NOT_EXIST),
            Optional.empty()));
  }
```
```
  // priv method 2
  private boolean nodeHasAccessToPrivacyGroup(final String privacyGroupIdString,
      final String enclaveKey) {
    // check whether the user (enclaveKey) is part of the current privacy group
    PrivacyGroup privacyGroup = null;
    try {
      privacyGroup = privacyParameters.getEnclave().retrievePrivacyGroup(privacyGroupIdString);
    } catch (final EnclaveClientException e) {
      // do not do anything. This is handled in the following if
    }
    if (privacyGroup == null || !privacyGroup.getMembers().contains(enclaveKey)) {
      return true;
    }
    return false;
  }
```

Reply:
Did some refactoring ....